Smooth Linear Advance with extruder synchronisation

dbuezas opened the pull request:
### Description

Introduces a new smooth linear advance method that decouples extrusion from motion constraints, improving print speed and quality without acceleration limits. (comparable to Klipper's approach).

* Smooth linear advance is managed by a new ISR that runs at a constant interval.
* Smoothing is calculated from planned extrusion rate instead of acceleration. This improves nozzle pressure control by making the algorithm aware of XY jerk and line width changes, and results in higher print quality compared to 'classic LA'.
* Instead of using moving window averaging (like Klipper) or instantaneous speed changes (like "classic" marlin linear advance), this PR uses an efficient approximation of gaussian filtering. The approximation consists of higher order (5) exponential smoothing corrected for delay.
* Delay correction is achieved by looking ahead in the block (even into future blocks). The smoothing factors are automatically calculated from the lookahead offset (TAU) to ensure delay compensation is precise and requires no calibration.
* TAU is somewhat analogous to Klipper's `pressure_advance_smooth_time` and can be configured in the LCD or via GCODE. The lower the better, as long as the extruder stepper doesn't skip steps.

Enabling SMOOTH_LIN_ADV results in faster and higher quality prints in 32 bit MCUs

The new ISR allows for a second second feature, `INPUT_SHAPING_E_SYNCH`, which synchronises the extruder with the input shaped x and y axes. This results in very accurate extrusion during acceleration phases. Synchronizing E to input shaping improves seams and top surfaces significantly, even at high accelerations (e.g nice seams and top surfaces at 15k mm/s^2).

### Requirements

32 bit MCU

### Benefits

Enables higher acceleration without extrusion artifacts
- Improves top surfaces and seam quality
- Reduces pressure variations in high-speed printing

### Configurations

```cpp
  #define SMOOTH_LIN_ADV          // Removes motion acceleration limitation by allowing to gradual ramp up of nozzle pressure
  #if ENABLED(SMOOTH_LIN_ADV)
    #define ADVANCE_TAU 0.01      // (secs) Smoothing time to reduce extruder acceleration due to linear advance.
                                  // It is also the time ahead the smoother needs to look in the planner
                                  // so the planner needs to have enough blocks loaded.
                                  // for k=0.04 at 10k acceleration and an orbiter 2, it can be as low as 0.0075.
                                  // The lower the better until stepper skipps
                                  // Higher k and higher print accelerations may require larger tau to avoid skipping steps
    #define SMOOTH_LIN_ADV_HZ 5000   // How often to update extruder speed
    #define INPUT_SHAPING_E_SYNCH    // Synchronize the extruder shaped xy axes (increrses precision)
  #endif

```

### Related Issues

This is a further evolution of- and closes #27352. Now with more solid mathematical foundations and improved precision.

### Misc

I have a fast 32 bit MCU with FPU so I made liberal use of floats. Further optimisations (e.g. fixed-point arithmetic, planner pre-computations) are possible. However, given the significant speed and quality improvements, I propose addressing those in a follow-up PR after merging this foundational implementation. I'm getting the speed and quality of a Voron or bambulab in my Ultimaker 2 with a direct drive extruder.

@tombrazier mentioned that, at least in theory, extrusion synchronisation could be done at step level. If implemented, it would replace my version of INPUT_SHAPING_E_SYNCH.

### Validation of this approach

Analysed stepper control signals with a logic analyzer and performed extensive testing (>1kg of prints)

Thanks @oliof and @tombrazier for the early feedback and discussions in discord.

dbuezas committed:
Smooth Linear Advance with extruder synchronisation

(cherry picked from commit 29720dce9cff7a8990bb7d7b0c72776ee79ee18a)

dbuezas committed:
keep shaper private parts private

(cherry picked from commit 28300f8cdff96399c5836cacb74c964da45fc453)

tombrazier commented:
First thought: what about a name change at this point? "Linear advance" isn't as good name as "pressure advance" and everyone knows "smooth pressure advance" from Klipper, so just the name would communicate a lot if we changed "linear" to "pressure". @thinkyhead

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2892,185 @@ hal_timer_t Stepper::block_phase_isr() {
+    }
+    #if ENABLED(INPUT_SHAPING_E_SYNCH)
+      constexpr uint16_t IS_COMPENSATION_BUFFER_SIZE = 
+        (SMOOTH_LIN_ADV_HZ / (20 /* min input shaping frequency*/) / 2.0f + 0.5f);
```
I think you need to use `SHAPING_MIN_FREQ` instead of 20.

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2892,185 @@ hal_timer_t Stepper::block_phase_isr() {
+      typedef struct {
+          float x;
+          float y;
+      } XYPair;
```
There is a general type `xy_float_t` for this.

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2892,185 @@ hal_timer_t Stepper::block_phase_isr() {
+      }
+
+      XYPair lookback(shaping_time_t t /* in stepper timer ticks */) {
+        constexpr float STEPPER_TICKS_PER_ADV_TICS = (float) SMOOTH_LIN_ADV_HZ / STEPPER_TIMER_RATE;
```
I think the variable name `STEPPER_TICKS_PER_ADV_TICS` is the wrong way round. Shouldn't it be `ADV_TICS_PER_STEPPER_TICKS`?

dbuezas replied:
good catch

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2892,185 @@ hal_timer_t Stepper::block_phase_isr() {
+      }
+      float total_rate = planned_step_rate + la_step_rate;
+
+      #if ENABLED(INPUT_SHAPING_E_SYNCH)
+        XYPair pre_shaping_rate = {0, 0},
+               first_pulse_rate = {0, 0};
+        float unshaped_rate_e = total_rate;
+        if (current_block) {
+          float xy_length = current_block->steps.x + current_block->steps.y;
+          if (xy_length > 0) {
+            unshaped_rate_e = 0;
+            pre_shaping_rate = {
+              total_rate * current_block->steps.x / xy_length,
+              total_rate * current_block->steps.y / xy_length,
+            };
+            first_pulse_rate = {
+              pre_shaping_rate.x * Stepper::shaping_x.factor1 / 128.0f,
+              pre_shaping_rate.y * Stepper::shaping_y.factor1 / 128.0f
+            };
+          }
+        }
+        const XYPair second_pulse_rate = {
+          lookback(ShapingQueue::get_delay_x()).x * Stepper::shaping_x.factor2 / 128.0f,
+          lookback(ShapingQueue::get_delay_y()).y * Stepper::shaping_y.factor2 / 128.0f
+        };
+        add_to_buffer(pre_shaping_rate);
+
+        const float x = first_pulse_rate.x + second_pulse_rate.x;
+        const float y = first_pulse_rate.y + second_pulse_rate.y;
+
+        total_rate = unshaped_rate_e + x + y;
+      #endif
```
Input shaping can be on any combination of X, Y and Z. Ideally this code should reflect that. (As should the related check `SanityCheck.h`.) However, as remarked at the top of the PR a better solution is probably to synchronise axes in IS, post the LA code.

dbuezas replied:
Yes, I avoided doing that because:
* I assume it is unusual not to have IS in both x and y (z may or may not be enabled)
* It also changes the lookback buffer's shape
* It will make the first PR on the topic harder to review
* and _"as remarked at the top of the PR a better solution is probably to synchronise axes in IS, post the LA code"_

tombrazier replied:
I am happy with cleaning this up as a separate PR in the IS code.

dbuezas committed:
tom's feedback

(cherry picked from commit 0f10d91b73d30c7f5fec71364ef0708aab896421)

dbuezas committed:
fix ambiguous assignment

(cherry picked from commit 2d74be5860748f5c72f2d71e45c5c1cfd015a96e)

tombrazier commented on Marlin/src/module/planner.h:
```diff
@@ -254,10 +265,14 @@ typedef struct PlannerBlock {
+    #if ENABLED(SMOOTH_LIN_ADV)
+      bool use_advance_lead;
+    #else
+      uint16_t max_adv_steps,                 // Max advance steps to get cruising speed pressure
+              final_adv_steps;               // Advance steps for exit speed pressure
+      uint8_t  la_scaling;                    // Scale ISR frequency down and step frequency up by 2 ^ la_scaling
+      uint32_t la_advance_rate;               // The rate at which steps are added whilst accelerating
```
This would be tidier if the code was an exact copy of the original except for the extra indentation. Would also fix the misalignment in the comments.

dbuezas replied:
ups...

dbuezas replied:
done: exact copy now

dbuezas committed:
Undo unnecessary variable reordering

dbuezas committed:
indentation

tombrazier commented on Marlin/src/module/planner.h:
```diff
@@ -1023,6 +1043,14 @@ class Planner {
+     *
+     * WARNING: Called from Stepper ISR context!
+     */
+    static block_t* lookahead(uint8_t offset);
```
I think `lookahead()` is a pretty generic name. Maybe `get_future_block()` instead?

dbuezas replied:
renamed to get_future_block

dbuezas committed:
undo empty line removal

tombrazier commented on Marlin/src/module/planner.h:
```diff
@@ -512,6 +527,11 @@ class Planner {
       static float extruder_advance_K[DISTINCT_E];
     #endif
+    #if ENABLED(SMOOTH_LIN_ADV)
+      static float extruder_advance_TAU,
+                   extruder_advance_TAU_TICKS,
+                   extruder_advance_ALPHA;
```
These variables are not used in the planner code. I think they should be moved to the stepper module.

dbuezas replied:
I had just put them next to `extruder_advance_K` without thinking about it.

Moved now

dbuezas committed:
rename Planner::lookahead with get_future_block

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -774,6 +779,21 @@ block_t* Planner::get_current_block() {
+  const uint8_t nr_moves = movesplanned();
+
+  // If there are any moves queued ...
+  if (nr_moves < offset) return nullptr;
```
Is there an off by one error here? If there was only the current block in the queue then I think `nr_moves` would be 1. And if `offset` were 1 then I think the next block is being requested, but there isn't one. So `nullptr` should be returned.

Should the condition not be `nr_moves <= offset`?

dbuezas replied:
```cpp
static inline uint8_t Planner::movesplanned()
Number of moves currently in the planner including the busy block, if any
```
You are right! good eye

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -774,6 +779,21 @@ block_t* Planner::get_current_block() {
 
   return nullptr;
 }
+block_t* Planner::lookahead(uint8_t offset) {
```
I think there's a potential race condition associated with this function. It looks ahead to blocks which still may be subject to change due to recalculation of the trapezoidal profiles. And since it is going to be called from within an ISR context the block's parameters may actually be in an inconsistent state when the ISR queries it.

dbuezas replied:
The inconsistent state should be avoided by the `if (block->flag.recalculate) return nullptr;` line.

You are right about one thing tho: since the fn doesn't mark the future block as busy, it is possible that the block ends up changing before it is executed, in which case the pressure target will have been wrong. This is not very problematic because:
* The smoothing filter will gracefully and smoothly start aiming at the corrected target without abrupt rate changes (no skipped steps)
* The lookahead necessary to keep extruder accel < 3k is typically around 10ms for direct drive, and 80ms for bowden, so only the beginning of the next block will be looked at.

tombrazier replied:
Oh, yeah, that makes sense. The whole thing is in ISR context including the `block->flag.recalculate` guard.

tombrazier replied:
To be honest this was the only thing I was worried about so far. Other comments have all been about code readability.

dbuezas committed:
move vars to Stepper

tombrazier commented on Marlin/src/module/planner.h:
```diff
@@ -240,11 +245,17 @@ typedef struct PlannerBlock {
 
-  #if ENABLED(S_CURVE_ACCELERATION)
+  #if ENABLED(SMOOTH_LIN_ADV)
+    uint32_t cruise_time;
```
Needs a comment. I think the correct comment would be "Cruise time in STEP timer counts".

dbuezas committed:
Fix off by one bug in get_future_block

(cherry picked from commit 4115faf52216c832e4562d1fc68d1ea524507a64)

dbuezas committed:
Fix after moving vars to Stepper::

(cherry picked from commit adb6c0087dc3bb210fcaa3fd58c3f5edd01e8ba1)

tombrazier commented on Marlin/src/module/planner.cpp:
```diff
@@ -2442,23 +2472,28 @@ bool Planner::_populate_block(
-      block->la_advance_rate = extruder_advance_K[E_INDEX_N(extruder)] * block->acceleration_steps_per_s2;
+    #if ENABLED(SMOOTH_LIN_ADV)
+      block->use_advance_lead = use_advance_lead;
+      block->xy_to_e_step_ratio = (block->direction_bits.e ? 1 : -1) * 
```
I found the name `xy_to_e_step_ratio` confusing. Partly because I read it backwards, i.e. xy to e ration would be xy / e in my head. Partly because xy may well not be x and/or y at all. Could even be e in fact. Maybe just `e_step_ratio` would be good.

dbuezas committed:
Comment cruise_time var

(cherry picked from commit 3722520134abb5a2a061027b415569847d192f47)

dbuezas committed:
xy_to_e_step_ratio => e_step_ratio

(cherry picked from commit 17ec816c0284f55c6fbc16cddcc16c49fa19a8d9)

dbuezas committed:
Smoothen total extrusion instead of only LA term (more silent extruder)

(cherry picked from commit 3b5a4a6baffa31e1f92d7ed93b7911bc8a15f1e4)

narno2202 commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2893,186 @@ hal_timer_t Stepper::block_phase_isr() {
+          count_direction.e = -count_direction.e;
+          DIR_WAIT_BEFORE();
+          E_APPLY_DIR(forward_e, false);
+          TERN_(FTM_OPTIMIZE_DIR_STATES, last_set_direction = last_direction_bits);
```
`FTM_OPTIMZE_DIR_STATES` is now always applied so the line must be : `TERN_(FT_MOTION, last_set_direction = last_direction_bits);`

dbuezas replied:
Oh, right, I had branched before. Fixed now, thank you.

dbuezas committed:
Fix missing semi colon

(cherry picked from commit 68624a514cd9f19034f3652948638197884ddaaa)

dbuezas committed:
fix last_set_direction

(cherry picked from commit ff053d207b78e4fd6ea0e20cfb151aeb6df71a50)

dbuezas committed:
Remove accidentally default def of LIN_ADVANCE

dbuezas committed:
Fix S_CURVE_ACCELERATION only macro

(cherry picked from commit c4e7fd0bc7aa6e6fda3c7344061a08bf4f67390c)

dbuezas committed:
Rename isr step countdown var

(cherry picked from commit b8de47c618367580377425f8a8a88e6fd9873b1f)

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2763,15 +2787,18 @@ hal_timer_t Stepper::block_phase_isr() {
-        la_active = (current_block->la_advance_rate != 0);
         #if DISABLED(MIXING_EXTRUDER) && E_STEPPERS > 1
           // If the now active extruder wasn't in use during the last move, its pressure is most likely gone.
           if (stepper_extruder != last_moved_extruder) la_advance_steps = 0;
```
`la_advance_steps` does not exist for `SMOOTH_LIN_ADV`.

dbuezas replied:
Wow, good catch. Fixed by putting the whole block under 
`#if ENABLED(LIN_ADVANCE) && DISABLED(SMOOTH_LIN_ADV)`

dbuezas replied:
For some reason github is not showing the block as outdated, but now line 2789 is 
`#if ENABLED(LIN_ADVANCE) && DISABLED(SMOOTH_LIN_ADV)`

tombrazier commented on Marlin/src/module/stepper.h:
```diff
@@ -348,6 +351,16 @@ class Stepper {
     #endif
 
+    #if ENABLED(SMOOTH_LIN_ADV)
+      static float  extruder_advance_TAU;       // The smoothing time, which is also the lookahead 
```
I would rather see `extruder_advance_TAU` declared as a private variable along with `extruder_advance_TAU_TICKS` and `extruder_advance_ALPHA` and then read from outside the `Stepper` class through an accessor function.

dbuezas replied:
Good call. Although it was safe to read, it wasn't safe to write directly

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2893,186 @@ hal_timer_t Stepper::block_phase_isr() {
+          Stepper::extruder_advance_TAU_TICKS,
+          Stepper::extruder_advance_ALPHA;
+
+    void Stepper::set_la_interval(int32_t rate) {
+      if (rate == 0) {
+        la_interval = LA_ADV_NEVER;
+      }
+      else {
+        const bool forward_e = rate > 0;
+        la_interval = calc_timer_interval(uint32_t(ABS(rate)));
+        if (forward_e != motor_direction(E_AXIS)) {
+          last_direction_bits.toggle(E_AXIS);
+          count_direction.e = -count_direction.e;
+          DIR_WAIT_BEFORE();
+          E_APPLY_DIR(forward_e, false);
+          TERN_(FT_MOTION, last_set_direction = last_direction_bits);
+          DIR_WAIT_AFTER();
+        }
+      }
+    }
```
This function can now be used to replace the code in `Stepper::block_phase_isr()` for non-smoothed linear advance. I think that would make cleaner code.

However the downside could be slightly less optimal code. I think if the function were forced to be inline it would help. Also adding a parameter called something line `bool canreverse` and using that to guard the direction reversing code might mean the inlining eliminates that block where it is not needed.

dbuezas replied:
Although I agree, I'm hesitant about this because: 
* The code just assigns la_advance without direction change in two places (during acceleration and cruise). There, set_la_interval would implicitly coerce from uint to int. I assume that will eat a couple cycles for non-smooth-la users
* During deacc, the rate may be negative, and the shift by la_scaling will result in a bug. To avoid this the code will get slightly more complex (i.e passing the shift amount as a parameter, then getting the absolute value of the rate inside set_la_interval, then shifting.), and there will be more total operations for non-smooth-la users.

Since these make the interrupt a bit slower, I'm afraid I'll be affecting performance for slow boards who can't use smooth advance. 

Maybe @thinkyhead would be best to make that call?

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2893,186 @@ hal_timer_t Stepper::block_phase_isr() {
+    
+    hal_timer_t Stepper::smooth_lin_adv_isr() {
+      static float soothed_values[SMOOTH_LIN_ADV_EXP_ORDER] = {0};
+      float target_pressure = 0;
```
The names of `target_pressure` and `last_target_pressure` confused me. They are not actually storing pressure but advanced steps. I think `target_adv_steps` or something similar would make more sense.

dbuezas replied:
You are right, renamed to target_adv_steps

tombrazier commented on Marlin/src/module/stepper.cpp:
```diff
@@ -2862,18 +2893,186 @@ hal_timer_t Stepper::block_phase_isr() {
+    };
+    
+    hal_timer_t Stepper::smooth_lin_adv_isr() {
+      static float soothed_values[SMOOTH_LIN_ADV_EXP_ORDER] = {0};
```
I suggest moving the declaration of `soothed_values` down to where it is used.

dbuezas replied:
done

tombrazier commented:
I have reviewed this entire PR. The whole thing makes sense to me and I think it does what it says on the tin. Testing results from @dbuezas look impressive. I think it should be merged.

The only thing I have not done is the maths to confirm is whether the value calculated for `extruder_advance_ALPHA` is correct to ensure that the smoothed extruder motion does not lead or lag. I take it that David is on top of the maths and, anyway, the test results appear to indicate that this is the case.

As @dbuezas says, there are probably future optimisations that can be added. e.g. possibly it can be reduced to fixed point maths so that more printers can access this logic. However it would be best to merge this now and focus on optimisations in the future.

tombrazier commented:
Just to add... I think this is a really nice piece of work.

dbuezas committed:
Maxes extruder_advance_tau private and adds accessors

(cherry picked from commit 9b5998d3dd34165c3d4962381cc76e4726d13389)

dbuezas committed:
Fixes "la_advance_steps does not exist for SMOOTH_LIN_ADV"

(cherry picked from commit 5c57e7033649b2fba441d252226fcd1ba577ab79)

dbuezas committed:
Renames target_pressure => target_adv_steps

(cherry picked from commit e25c0198ca681164e29526b31c8b5d14fc4de695)

dbuezas commented:
Thanks for the extensive review @tombrazier !

> The only thing I have not done is the maths to confirm is whether the value calculated for `extruder_advance_ALPHA` is correct to ensure that the smoothed extruder motion does not lead or lag. I take it that David is on top of the maths and, anyway, the test results appear to indicate that this is the case.

This relationship is `α=1−exp(−dt/τ)` here's a source 
(second equation under **Exponential filter equations and initialization**): 
https://gregstanleyandassociates.com/whitepapers/FaultDiagnosis/Filtering/Exponential-Filter/exponential-filter.htm

In this PR's case, I am chaining 5 smoothers to better approximate a gaussian filter, so I divide tau by 5 (`SMOOTH_LIN_ADV_EXP_ORDER is 5`), to get the alpha (smoothing constant) of each filter. This means each filter pass will delay the signal by `tau/5`, totalling `tau` delay. 

I also "experimentally" confirmed the delay correction is correct through multiple simulations with varying values of tau and different accelerations. Here's one of those simulations:
```cpp
k = 0.04
tau = 0.01
accel = 15000
jerk = 20
vel = 300
```
### Without delay correction
<img width="1235" alt="image" src="https://github.com/user-attachments/assets/33ac3468-610f-4dd7-b4e4-c69c8d51db63" />

### Delay corrected (with lookahead of tau)
<img width="1236" alt="image" src="https://github.com/user-attachments/assets/3d41d779-bafc-4c42-8dbc-9f730fd5268e" />

tombrazier commented:
Thanks for the maths explanation. It's the sort of thing I go looking for the PRs when I want to understand some code.

dbuezas commented:
I should add that this phase delay will be wrong when the input signal's period is too close to tau (i.e the total duration of a block is close to tau).
But error doesn't ever accumulate since I'm tracking the target advance steps instead of rate, and the amount of advance steps in a 10ms block is negligible anyway, so this limit is not a problem at all.

dbuezas committed:
Don't smooth planned extrusion since that's not delay compensated

(cherry picked from commit b41e8634b4d53ea3a9ba894b7f8fd1252bac6797)

dbuezas committed:
Do lookahead during unretractions

This may improve high speed seams a small bit but it is very dependant on machine properties and can result in underextrusion too

(cherry picked from commit adc7ff273dd2b33a6364d26bd949e0f5a130e203)

dbuezas committed:
correct indentation

(cherry picked from commit 2fb6c6b28cef48a0414f1479564f01cdee04d0d6)

classicrocker883 commented:
### Here are some changes I propose:

in multiple files:
```diff
- #if (DISABLED(SMOOTH_LIN_ADV))
+ #if DISABLED(SMOOTH_LIN_ADV)

- #if (ENABLED(SMOOTH_LIN_ADV))
+ #if ENABLED(SMOOTH_LIN_ADV)
```

<br>

to be like **stepper.cpp**,
in **stepper.h** line ~450
```diff
-       static uint32_t curr_timer_tick, // Current tick relative to block start
-                       curr_step_rate;  // Current motion step rate

+       static uint32_t curr_step_rate,  // Current tick relative to block start
+                       curr_timer_tick; // Current motion step rate
```

<br>


So I know `S_CURVE_ACCELERATION` with `SMOOTH_LIN_ADV` may not work together, however when both are enabled there is an error, to fix this...
in **planner.cpp** line ~2500
```diff
- #if DISABLED(S_CURVE_ACCELERATION)
    block->acceleration_rate = (uint32_t)(accel * (float(1UL << 24) / (STEPPER_TIMER_RATE)));
  #endif

+ #if DISABLED(S_CURVE_ACCELERATION) || ENABLED(SMOOTH_LIN_ADV)
    block->acceleration_rate = (uint32_t)(accel * (float(1UL << 24) / (STEPPER_TIMER_RATE)));
  #endif
```

in **planner.h** line ~260
```diff
  #if ENABLED(S_CURVE_ACCELERATION)
  ...
- #else
    uint32_t acceleration_rate;             // Acceleration rate in (2^24 steps)/timer_ticks*s
  #endif

  #if ENABLED(S_CURVE_ACCELERATION)
  ...
+ #endif
+ #if DISABLED(S_CURVE_ACCELERATION) || ENABLED(SMOOTH_LIN_ADV)
    uint32_t acceleration_rate;             // Acceleration rate in (2^24 steps)/timer_ticks*s
  #endif
```

thinkyhead committed:
quick search/replace

thinkyhead committed:
initial review changes

thinkyhead committed:
"SMOOTH_LIN_ADVANCE"

thinkyhead committed:
add test

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/27710

thinkyhead commented:
A couple of things I'd like to adjust before completion, and already did a little work on:
- Implement support for `DISTINCT_E_FACTORS`.
- Allow `INPUT_SHAPING_E_SYNC` to be enabled when only one of `INPUT_SHAPING_X` or `INPUT_SHAPING_Y` is enabled.

thinkyhead committed:
etc

thinkyhead committed:
more etc

thinkyhead committed:
towards distinct e

thinkyhead committed:
fix build errors, warnings

thinkyhead committed:
and so on

thinkyhead committed:
formatting

thinkyhead committed:
shorthand

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27710

thinkyhead committed:
outdent

dbuezas commented:
Another future improvement is to make it follow the s curve profile, although I think s curve lost relevance with the introduction of IS

thinkyhead committed:
misc.

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27710

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27710

thinkyhead committed:
Test for INPUT_SHAPING_E_SYNC

thinkyhead commented:
The place where S-curve still comes in handy is with non-printing moves. It might make sense to have S-curve only apply to non-printing moves when it and linear advance are both enabled — to remove any artifacts caused by their interactions, though I think they would be minimal on small features. Some time we'll have to take some samples to make a graph and see how they interact.

dbuezas commented:
Wow, easter surprise! I'll retest the merged version as soon as I have a chance early next week

dbuezas commented:
Retested, all in working order. Thanks @thinkyhead !

vovodroid commented:
@thinkyhead 

> It might make sense to have S-curve only apply to non-printing moves when it and linear advance are both enabled

Are you referring to "Classic" LA or new one?

dbuezas commented:
Referring to smooth advance. "Classic"/"rough"  linear advance cannot follow the s curve that precisely.
The new fixed frequency isr of smooth advance can do that, it just need to be able to compute the future speed value (basically reuse what already exists for the s-curve)

It will make advance even smoother, but I don't know of people use both input shaping and s curve. Not sure if it is even worth enabling both at the same time)

oliof commented:
ISTR that s curve can double the step rate, but I don't remember if that only applies to 8bit boards or not (don't see why it should). So even if no one did it today, it may still be a useful combination for improving the speed of prints, especially since IS would also mostly be relevant for printing corners while S curve would beneficiary for (longer) non printing moves in thinkyheads proposal

Gesendet von Proton Mail für Android

-------- Ursprüngliche Nachricht --------
Am 22.04.25 14:51 schrieb David Buezas :

> Referring to smooth advance. "Classic"/"rough" linear advance cannot follow the s curve that precisely.
> The new fixed frequency isr of smooth advance can do that, it just need to be able to compute the future speed value (basically reuse what already exists for the s-curve)
>
> It will make advance even smoother, but I don't know of people use both input shaping and s curve. Not sure if it is even worth enabling both at the same time)
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/MarlinFirmware/Marlin/pull/27710#issuecomment-2821231962), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAJMOIZH3F7Z6GUKX25LUJ322Y3M3AVCNFSM6AAAAABXYZBFMOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRRGIZTCOJWGI).
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
> dbuezas left a comment [(MarlinFirmware/Marlin#27710)](https://github.com/MarlinFirmware/Marlin/pull/27710#issuecomment-2821231962)
>
> Referring to smooth advance. "Classic"/"rough" linear advance cannot follow the s curve that precisely.
> The new fixed frequency isr of smooth advance can do that, it just need to be able to compute the future speed value (basically reuse what already exists for the s-curve)
>
> It will make advance even smoother, but I don't know of people use both input shaping and s curve. Not sure if it is even worth enabling both at the same time)
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/MarlinFirmware/Marlin/pull/27710#issuecomment-2821231962), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AAJMOIZH3F7Z6GUKX25LUJ322Y3M3AVCNFSM6AAAAABXYZBFMOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRRGIZTCOJWGI).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

vovodroid commented:
@dbuezas did you notice vibration increase after disabling s-curve?

dbuezas commented:
I never really used s-curve. Only tried it once shortly

vovodroid commented:
Something like this:

```
    void Stepper::set_la_interval(int32_t step_rate) {
      if (step_rate == 0 || !ANY_AXIS_MOVES(current_block)) {
        la_interval = LA_ADV_NEVER;
      }
```

dbuezas commented:
Smooth linear advance could still be applying corrections after the other axes stopped, the smoothing makes it start accelerating tau seconds before planned, and stop (roughly) tau seconds after. This is by design to avoid discontinuities is rate.

Also set_la_interval actually covers all extrusion, not only the linear advance part. 


Regarding the error you observe, it may be numerical error since it works on rates and not positions, but 5% is too high to be explained by that. Do you also get that kind of variance with NLE disabled?

dbuezas commented:
Mmm maybe, hard to say. A very low tau would make it act like the old LA

vovodroid commented:
Even with zero LA?

vovodroid commented:
> The difference is that smooth LA changes the extruder speed 1000 times a second, instead of exactly in synch with the planner execution

Should it apply correction when extrusion speed is constant?

vovodroid commented:
Well, I build with following config without NLE:

```
#define LIN_ADVANCE
#define ADVANCE_K 0.22        // (mm) Compression length applying to all extruders
#define SMOOTH_LIN_ADVANCE    // Remove limits on acceleration by gradual increase of nozzle pressure
#define ADVANCE_TAU 0.02       // (s) Smoothing time to reduce extruder acceleration
#define SMOOTH_LIN_ADV_HZ 1000   // (Hz) How often to update extruder speed
#define INPUT_SHAPING_E_SYNC     // Synchronize the extruder-shaped XY axes (to increase precision)
//#define NONLINEAR_EXTRUSION
```

Here is underextrusion table for speed (mm/s):

```
3	0.929
6	0.926
18	0.985
30	0.968
39	0.944
```

Here is a graph:

![image](https://github.com/user-attachments/assets/cd20d8da-608a-4227-937d-1aa74c155a28)

No LA/Tau change affect this.

vovodroid commented:
Just enable LA_DEBUG?

vovodroid commented:
Ahh, I checked it - without SMOOTH_LIN_ADVANCE it's always the same correct length.

dbuezas commented:
Ok that's super weird. With so much under extrusion it would really show

dbuezas commented:
But quite noticeable onTm top surfaces

vovodroid commented:
It depends on the speed so. Check it in your printer, it's very simple check.

vovodroid commented:
> track actual steps done

Sorry?

> leave the non LA term be handled directly by the block interrupt as before

Use smoothing only for acceleration/deceleration phase?

dbuezas commented:
> getting alot of skipped steps.

Make sure to keep print acceleration within reasonable limits and Increase tau. K=1.08 puts a lot of stress on the extruder
